crates/services/p2p/src/heavy_task_processor.rs

```diff
@@ -0,0 +1,189 @@
+use std::sync::Arc;
+use tokio::sync::Semaphore;
+
+pub struct HeavyTaskProcessor {
+    rayon_thread_pool: rayon::ThreadPool,
+    semaphore: Arc<Semaphore>,
+}
+
+impl HeavyTaskProcessor {
+    pub fn new(
+        number_of_threads: usize,
+        number_of_pending_tasks: usize,
+    ) -> anyhow::Result<Self> {
+        let rayon_thread_pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(number_of_threads)
+            .build()
+            .map_err(|e| anyhow::anyhow!("Failed to create a rayon pool: {}", e))?;
+        let semaphore = Arc::new(Semaphore::new(number_of_pending_tasks));
+        Ok(Self {
+            rayon_thread_pool,
+            semaphore,
+        })
+    }
+
+    pub fn spawn<OP>(&self, op: OP) -> bool
```

Comment:
Instead of `bool`, I'd like to use `Result<(), OutOfCapacity>` or something. This way it's not possible to forget to check the return value, as that would trigger `unused_must_use` lint.